presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PropertyDerivations.java

```diff
@@ -449,9 +449,12 @@ public ActualProperties visitJoin(JoinNode node, List<ActualProperties> inputPro
                                 .build();
                     }
 
-                    if (isOptimizeFullOuterJoinWithCoalesce(session) && probeProperties.getNodePartitioning().isPresent() && buildProperties.getNodePartitioning().isPresent()) {
+                    if (isOptimizeFullOuterJoinWithCoalesce(session) &&
+                            probeProperties.getNodePartitioning().isPresent() &&
+                            buildProperties.getNodePartitioning().isPresent() &&
+                            metadata.getCommonPartitioning(session, probeProperties.getNodePartitioning().get().getHandle(), buildProperties.getNodePartitioning().get().getHandle()).isPresent()) {
```

Comment:
what about 
```
   && metadata.isRefinedPartitioningOver(session, probeProperties.getNodePartitioning().get().getHandle(), buildProperties.getNodePartitioning().get().getHandle()).isPresent()) || metadata.isRefinedPartitioningOver()...
```

You might want to save `probeProperties.getNodePartitioning().get().getHandle()` and `buildProperties.getNodePartitioning().get().getHandle()).isPresent()` as local variable.